File: cat2/wekker/ruby/woumpousse.rb
def shortest_hours(h, hh)
  [ [h, hh].max - [h, hh].min, [h, hh].min - [h, hh].max + 24 ].min
end

def solve(h, m, hh, mm)
  if m <= mm
  then
    solutions = []
  
    solutions << mm - m + shortest_hours(h, hh)
    solutions << 60 + m - mm + shortest_hours((h - 1 + 24) % 24, hh)

#    puts " #{[h,m,hh,mm]} : #{solutions}"

    solutions.min
  else
    solve(hh, mm, h, m)
  end
end


def main
  (1..gets.to_i).each do |index|
    h, m = gets.split(/ /).map(&:to_i)
    hh, mm = gets.split(/ /).map(&:to_i)

    puts "#{index} #{solve(h,m,hh,mm)}"
  end
end


main

File: cat4/delphi/python/woumpousse.py
class Solver:
    def __init__(self):
        self.__table = {}
        
    def solve(self, k, n):
        if k == 1:
            return n
        elif n == 0:
            return 0
        else:
            if not (k, n) in self.__table:
                self.__table[(k, n)] = min( max([self.solve(k-1, i-1), 1 + self.solve(k, n-i)]) for i in range(1, n+1) )

            return self.__table[(k, n)]


n = int(input())
solver = Solver()

for index in range(1, n+1):
    k, n = map(int, input().split(' '))
    solution = solver.solve(k, n)

    print(index, solution)
        
